ggml/src/ggml-fp8.cpp

```diff
@@ -0,0 +1,279 @@
+#include <cassert>
+#include <algorithm>
+
+#define GGML_COMMON_DECL_CPP
+#include "ggml-common.h"
+#include "ggml.h"
+
+#include "ggml-fp8.h"
+
+/*
+make clean
+make -j8
+# ./llama-quantize --output-tensor-type fp8_e3m4_q ~/LLM/Mistral-Nemo-Instruct-2407.BF16.gguf ~/LLM/Mistral-Nemo-Instruct-2407.E3M4_Q.gguf E3M4_Q
+./llama-quantize ~/LLM/Mistral-Nemo-Instruct-2407.BF16.gguf ~/LLM/Mistral-Nemo-Instruct-2407.E3M4_Q.gguf E3M4_Q
+./llama-cli -c 1024 -m ~/LLM/Mistral-Nemo-Instruct-2407.E3M4_Q.gguf -p "[INST]bonjour a tu un nom. je ne sais pas comment t'appeler. Si tu n'en as pas je peux t'appeler TINTIN[/INST]" -s 42
+./llama-perplexity --kl-divergence-base ~/LLM/Mistral-Nemo-Instruct-2407.BF16.kld --kl-divergence -s 31337 -m ~/LLM/Mistral-Nemo-Instruct-2407.E3M4_Q.gguf
+
+rm -rf build
+cmake -B build
+cmake --build build --config Release -j $(nproc)
+./build/bin/llama-quantize ~/LLM/Mistral-Nemo-Instruct-2407.BF16.gguf ~/LLM/Mistral-Nemo-Instruct-2407.E3M4_Q.gguf E3M4_Q
+./build/bin/llama-cli -c 1024 -m ~/LLM/Mistral-Nemo-Instruct-2407.E3M4_Q.gguf -p "[INST]bonjour a tu un nom. je ne sais pas comment t'appeler. Si tu n'en as pas je peux t'appeler TINTIN[/INST]" -s 42
+./build/bin/llama-perplexity --kl-divergence-base ~/LLM/Mistral-Nemo-Instruct-2407.BF16.kld --kl-divergence -s 31337 -m ~/LLM/Mistral-Nemo-Instruct-2407.E3M4_Q.gguf
+
+# la CI local:
+rm -rf tmp
+mkdir tmp
+bash ./ci/run.sh ./tmp/results ./tmp/mnt
+
+# HIP legacy target?
+cmake -B build2 -S . -DCMAKE_C_COMPILER=hipcc -DCMAKE_CXX_COMPILER=hipcc -DGGML_HIPBLAS=ON
+cmake --build build2 --config Release -j $(nproc) -v
+
+*/
+
+template<int N> constexpr float EXP2() {
+    if constexpr (N==0) return 1;
+    if constexpr (N>0) return EXP2<N-1>()*2;
+    if constexpr (N<0) return EXP2<N+1>()/2;
+}
+
+// 2^N avec N>0 en entier
+template<int N> constexpr int EXP_I2() {
+    if constexpr (N==0) return 1;
+    if constexpr (N>0) return EXP_I2<N-1>()*2;
+}
+
+template<int _E> //, int M=7-E>  1.7 bits!
+struct FP8 {
+    uint8_t bits;
+    using type = FP8<_E>;
+    static constexpr int E=_E;
+    static constexpr int M=7-_E;
+    static constexpr int E_BIAS=EXP2<_E-1>()-1;
```

Comment:
The not c++11 support is this 'static constexpr"

Reply:
`static const int` would do the same here.